### ISSUE TITLE
[question] The generated JWT token can be decode, is this correct and secure?

### ISSUE THREAD

**Issue by laoshanxi** [Thalhammer/jwt-cpp#145]
Dear,

I found a URL from Microsoft which can decode any JWT token, so is the token secure enough to used for authentication?
![image](https://user-images.githubusercontent.com/50568160/123384963-9fbb6a00-d5c7-11eb-8145-6bd7eee77565.png)

https://adfshelp.microsoft.com/jwtdecoder/gettoken

**Comment by Thalhammer** [comment]
Yes.

JWT in it's raw form stores the data you provide as raw text. There is no encryption of any sort. 
It does however verify its integrity using a hash, that's cryptographically signed (the part after the last dot). This means that you can read what's inside a jwt, but you can't change it without making the token invalid. So don't store anything private inside the token (e.g. the users password) cause everyone having the token can read it (even after the token expired). However there's also no need to do this anyway. If you get a token and it correctly verifies, you are guaranteed that only someone with the private key could have created it (I.e. you). There is no way (assuming you used a secure algorithm and never leaked the private part of your key) to fake that token.

TLDR:
Are they safe for authentication: **Yes**
Are they encrypted/private: *No(*)**

You can do the same thing on jwt's official site as well BTW.

* There are extensions for encrypting the token, but we don't support them and it does not seem like there's any widespread use of them.

**Comment by laoshanxi** [comment]
Thanks for the answer.

So from client side, it is also necessary to keep the token in a secure place for cache usage?

**Comment by Thalhammer** [comment]
Depends on your application. Usually the one that has the password to the account, is also the one that has the token. There's no point in hiding anything from them since it's their data in the first place and again, there is no reason to store **anything** that's not information the user already knows. If you can, it's probably a good idea, but there's usually no reason to do so.

**Comment by laoshanxi** [comment]
> Depends on your application. Usually the one that has the password to the account, is also the one that has the token. There's no point in hiding anything from them since it's their data in the first place and again, there is no reason to store **anything** that's not information the user already knows. If you can, it's probably a good idea, but there's usually no reason to do so.

I use JWT also for a command line authentication by a REST Server, so I need cache the token on local host for a while:
https://app-mesh.readthedocs.io/en/latest/CLI.html#command-lines
